pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -413,9 +413,23 @@ func buildAutoMtlsSettings(
 	meshExternal bool,
 	serviceMTLSMode model.MutualTLSMode) (*networking.ClientTLSSettings, mtlsContextType) {
 	if tls != nil {
-		// Update TLS settings for ISTIO_MUTUAL.
-		// Use client provided SNI if set. Otherwise, overwrite with the auto generated SNI.
-		// user specified SNIs in the istio mtls settings are useful when routing via gateways.
+		if tls.Mode == networking.ClientTLSSettings_DISABLE || tls.Mode == networking.ClientTLSSettings_SIMPLE {
+			return tls, userSupplied
+		}
+		// For backward compatibility, use metadata certs if provided.
+		if hasMetadataCerts(proxy.Metadata) {
+			// When building Mutual TLS settings, we should always user supplied SubjectAltNames and SNI
+			// in destination rule. The Service Accounts and auto computed SNI should only be used for
+			// ISTIO_MUTUAL.
+			return buildMutualTLS(tls.SubjectAltNames, tls.Sni, proxy), userSupplied
+		}
+		if tls.Mode != networking.ClientTLSSettings_ISTIO_MUTUAL {
```

Comment:
I still don't understand why the SNI block is after. What is the case where tls.Mode is not ISTIO_MUTUAL ? Regular mTLS ? Either way, it should probably include the SNI.

Reply:
Yes. For regular mTLS if user has not specified SNI in TLS - are you saying we should still apply SNI that we compute?

Reply:
No, you are right, never mind.